File: src/main/java/at/ac/tuwien/complang/recursive/OptimizedNQueensSolverRecursive.java
package at.ac.tuwien.complang.recursive;

/**
 * Based on OriginalNQueensSolver but uses bit fields instead of arrays.
 */
public class OptimizedNQueensSolverRecursive {

    private final int n;
    private final int mid;

    // 1..n
    // a[x] ist true, wenn man in Spalte x eine Dame setzen kann
    private long aLong;

    // 2..2*n
    // b[x+y] ist true, wenn man in der Diagonale, 
    // die die Felder (x+k, y-k) (fuer alle k) enthaelt, eine Dame setzen kann
    private long bLong;

    // 1..2*n-1
    // c[x-y+n] ist true, wenn man in der Diagonale,
    // die die Felder (x+k, y+k) (fuer alle k) enthaelt, eine Dame setzen kann
    private long cLong;

    private long count = 0;

    public OptimizedNQueensSolverRecursive(int n) {
        this.n = n;
        this.mid = n >> 1;
        // All bits 1 except at index 0
        this.aLong = -2L;
        // All bits 1 except at index 0 and 1
        this.bLong = -4L;
        // All bits 1 except at index 0
        this.cLong = -2L;
    }
    
    private boolean a(int index) {
        return get(aLong, index);
    }
    
    private void a(int index, boolean value) {
        aLong = set(aLong, index, value);
    }
    
    private boolean b(int index) {
        return get(bLong, index);
    }
    
    private void b(int index, boolean value) {
        bLong = set(bLong, index, value);
    }
    
    private boolean c(int index) {
        return get(cLong, index);
    }
    
    private void c(int index, boolean value) {
        cLong = set(cLong, index, value);
    }
    
    private boolean get(long field, int index) {
        return (field & (1 << index)) != 0;
    }
    
    private long set(long field, int index, boolean value) {
        if (value) {
            return field | (1 << index);
        } else {
            return field & ~(1 << index);
        }
    }
    
    private void tryMid(final int i, final int j) {
        int bIndex = i + j;
        int cIndex = i + n - j;
        
        if (a(j) && b(bIndex) && c(cIndex)) {
            a(j, false);
            b(bIndex, false);
            c(cIndex, false);
            try1(i + 1, true);
            a(j, true);
            b(bIndex, true);
            c(cIndex, true);
        }
    }

    private void try1(final int i, boolean mirrorExit) {
        int j = 0;
        int bIndex = i;
        int cIndex = i + n;
        do {
            j++;
            bIndex++;
            cIndex--;
            if (mirrorExit && j > mid) {
                // Don't count mirrored solutions
                return;
            }
            if (a(j) && b(bIndex) && c(cIndex)) {
                if (i < n) {
                    a(j, false);
                    b(bIndex, false);
                    c(cIndex, false);
                    try1(i + 1, false);
                    a(j, true);
                    b(bIndex, true);
                    c(cIndex, true);
                } else {
                    count++;
                }
            }
        } while (j != n);
    }

    public long solve() {
        try1(1, true);
        
        if ((1 & n) != 0) {
            tryMid(1, mid + 1);
        }
        
        // Add mirrored solutions
        count = count << 1;
        
        return count;
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        OptimizedNQueensSolverRecursive o = new OptimizedNQueensSolverRecursive(n);
        System.out.println(o.solve());
    }
}

File: src/main/c/OptimizedNQueensSolverRecursive.c
/******************************************************************************
 * Based on OriginalNQueensSolver.
 *
 * Removed field x since there is no real use for it
 * Removed maxN and used n instead (use malloc to reserve int arrays on heap)
 * Common subexpressions eliminated
 * Symmetrie checking along the vertical middle line
 * Use long number and bitmask as boolean array (n limited to 31 now)
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

/* this is the actual board size to be used */
int n;
/* half of the board size */
int mid;
/* number of found solutions to the n-queens problem */
long count = 0L;
/* a[x] is true, if a queen can be placed in column x  (x: 1..n)
 We use here a int number and bitmask instead of  int a[n + 1] */
int aInt = -2; // all bits 1 except at index 0
/* b[y+x] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y-k) for all k. "/" antidiagonal line
 We use here a long number and bitmask instead of  int b[2 * n + 1] */
long bLong = -4L; // all bits 1 except at index 0 and 1
/* c[y-x+n] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y+k) for all k. "\" diagonal line
 We use here a long number and bitmask instead of  int c[2 * n] */
long cLong = -2L; // all bits 1 except at index 0

/* Function prototypes */
void try1(int i, int mirrorExit);
void tryMid(int i, int j);

/*
 * If we have odd number of n, then we have to do an extra run for the n/2 + 1
 * column.
 */
void tryMid(int i, int j) {
	// Common subexpressions
	int aBitMask = 1 << j;
	long bBitMask = 1L << (i + j);
	long cBitMask = 1L << (i - j + n);
	if ((aInt & aBitMask) && (bLong & bBitMask) && (cLong & cBitMask)) {
		aInt = aInt & ~aBitMask;
		bLong = bLong & ~bBitMask;
		cLong = cLong & ~cBitMask;
		try1(i + 1, 1);
		aInt = aInt | aBitMask;
		bLong = bLong | bBitMask;
		cLong = cLong | cBitMask;
	}
}

/*
 * Search all solutions that start with a queen in the row i. If mirrorExit,
 * then we do only columns 1..n/2 . With mirrorExit we control that only 1..n/2
 * is checked on the first row, for all deeper rows we do as usual 1..n.
 */
void try1(int i, int mirrorExit) {
	int j = 0;
	// Common subexpressions
	int bIndex = i;
	int cIndex = i + n;
	do {
		j++;
		if (mirrorExit && (j > mid)) {
			// Do not count mirrored solutions.
			return;
		}
		// Common subexpressions
		bIndex++;
		cIndex--;
		int aBitMask = 1 << j;
		long bBitMask = 1L << (bIndex);
		long cBitMask = 1L << (cIndex);
		if ((aInt & aBitMask) && (bLong & bBitMask) && (cLong & cBitMask)) {
			aInt = aInt & ~aBitMask;
			bLong = bLong & ~bBitMask;
			cLong = cLong & ~cBitMask;
			if (i < n) {
				try1(i + 1, 0);
			} else {
				count++;
			}
			aInt = aInt | aBitMask;
			bLong = bLong | bBitMask;
			cLong = cLong | cBitMask;
		}
	} while (j != n);
}

int main(int argc, char *argv[]) {
	if (argc <= 1 || (n = atoi(argv[1])) <= 0) {
		n = 8;
	}
	mid = n >> 1; // mid = n/2
	try1(1, 1);
	// if we have odd n, we have to do column n/2 + 1 extra
	if ((1 & n) == 1) {
		tryMid(1, (mid + 1));
	}
	// add mirrored solutions (simply double count)
	count = count << 1;
	printf("%ld\n", count);
	return 0;
}

